src/client/meshgen/collector.cpp

```diff
@@ -95,10 +96,47 @@ PreMeshBuffer &MeshCollector::findBuffer(
 	if (numVertices > U16_MAX)
 		throw std::invalid_argument(
 				"Mesh can't contain more than 65536 vertices");
+
 	std::vector<PreMeshBuffer> &buffers = prebuffers[layernum];
+
+	// avoid scanning the entire list of buffers when filling with the same material
+	if (latest_buffers[layernum].size() > 0 && 
+			buffers[latest_buffers[layernum].back()].layer == layer && 
+			buffers[latest_buffers[layernum].back()].vertices.size() + numVertices <= U16_MAX)
+		return buffers[latest_buffers[layernum].back()];
+
 	for (PreMeshBuffer &p : buffers)
-		if (p.layer == layer && p.vertices.size() + numVertices <= U16_MAX)
+		if (!p.closed && p.layer == layer && p.vertices.size() + numVertices <= U16_MAX)
 			return p;
+
 	buffers.emplace_back(layer);
+	latest_buffers[layernum].push_back(buffers.size()-1);
 	return buffers.back();
 }
+
+void MeshCollector::startNewMeshLayer(bool allow_reuse)
+{
+	for (s16 tile_layer = 0; tile_layer < MAX_TILE_LAYERS; ++tile_layer) {
+		// do not close mesh buffer if there is only one
+		if (latest_buffers[tile_layer].size() == 1)
+			continue;
+
+		s16 latest_buffer = -1;
+		if (allow_reuse && latest_buffers[tile_layer].size() > 0)
+			latest_buffer = latest_buffers[tile_layer].back();
+
+		for (s16 index: latest_buffers[tile_layer])
```

Comment:
This code should have comments to explain its' purpose and use.

Reply:
I've now simplified the method and added comments.